service/history/workflow/mutable_state_impl.go

```diff
@@ -3589,6 +3589,9 @@ func (ms *MutableStateImpl) ApplyWorkflowExecutionUpdateCompletedEvent(
 		}
 		sizeDelta = ui.Size() - sizeBefore
 	} else {
+		// TODO (alex): this should never happened because UpdateInfo should always be created before
```

Comment:
If something could only happen due to a bug, the best way is to panic. If this could happen due to a user/system error then you should return an error.

Reply:
I will think about it when address this TODO. But all panics on the API call code path (like this one) are captured and converted to internal errors which is returned to the client. Returning specific error makes more sense here.